Share a feature extractor between classifiers

vinvinod opened the pull request:
This is an attempt to share same feature extractor between classifiers.
Modified only named_entity_extractor and text_categorizer.

davisking commented:
Cool, thanks for the PR.  I'll take a look at it this week.  

On first glance though, I see there is a mixture of tabs and spaces in the change.  Can you update it so it's only spaces, this way the code looks uniform and the github change view isn't messed up?

vinvinod commented:
I have converted all tabs to spaces.

davisking commented:
I still see a bunch of tabs and bad indenting. :(

vinvinod commented:
I am really sorry! I should have been thorough.
Hopefully, now I haven't missed anything.

davisking commented:
No worries.

davisking commented on mitielib/include/mitie.h:
```diff
@@ -151,6 +151,13 @@ extern "C"
+        const char* filename
+    );
+    /*!
+	Added for Rasa_nlu multitenancy
```
What does this mean?

More generally, all the functions in this file should be properly documented.  e.g. what does this function do?  What are the requirements for calling it?

davisking commented on examples/python/text_categorizer_multiple_models.py:
```diff
@@ -0,0 +1,66 @@
+
+text_feature_extractor = total_word_feature_extractor(fe_filename)
+
+cat1_new = text_categorizer("new_text_categorizer_1_pure_model.dat", None, True)
```
The user shouldn't have to supply None,True as an argument.  The code can figure out that the file on disk is a pure model so it should not require the user to supply this redundant information.

davisking commented on mitielib/include/mitie/total_word_feature_extractor.h:
```diff
@@ -128,7 +128,7 @@ namespace mitie
                 return;
             }
-
+            dlib::matrix<float,0,1> temp;
```
Why is this here?

davisking commented on mitielib/include/mitie/word_morphology_feature_extractor.h:
```diff
@@ -64,6 +64,7 @@ namespace mitie
                 - #feats.size() == get_num_dimensions()
         !*/
         {
+            std::vector<dlib::uint16> hits;
```
Why is this here?

davisking commented on mitielib/include/mitie/word_morphology_feature_extractor.h:
```diff
@@ -93,6 +94,7 @@ namespace mitie
                   iterator range.
         !*/
         {
+            std::vector<dlib::uint16> hits;
```
Why is this here?  

This is the third time I've seen variables like this added that just alias existing variables.  Please don't do this, unless there is some very good reason to do it (which is not apparent to me).

davisking commented on mitielib/src/text_categorizer.cpp:
```diff
@@ -71,13 +71,23 @@ namespace mitie
+        string& text_tag,
+        double& text_score,
+        const total_word_feature_extractor& fe_
+    ) const {
```
Try to use the same formatting as the rest of the code, i.e. put { on it's own line.

davisking commented on mitielib/src/text_categorizer.cpp:
```diff
@@ -71,13 +71,23 @@ namespace mitie
             temp = df.predict(extract_BoW_features(sentence));
         } else {
-            const std::vector<matrix<float, 0, 1> > &sent = sentence_to_feats(fe, sentence);
+            const std::vector<matrix<float, 0, 1> > &sent = sentence_to_feats(fe_, sentence);
```
Why is fe being renamed to fe_?

vinvinod commented:
Hi Davis,

Added the required documentation for the functions and examples. 
Some redundant arguments have been removed as you have pointed out.
Also, I have removed the unnecessary temp variables I was using at some places.

davisking commented:
Super, I'll look it over in a few days.

davisking commented on examples/python/text_categorizers_with_shared_feature_extractor.py:
```diff
@@ -0,0 +1,77 @@
+    # Observe that we are passing the same feature extractor to both the categorizers
+
+    text1 = "I am so happy"
+    pred, conf = cat1_new(tokenize(text1), text_feature_extractor)
```
I've looked over the PR and everything looks great, except for one more minor detail.  I noticed that if I call this without giving a feature extractor the code executes and silently does the wrong thing.  Can you make it throw an error if the user tries to make a prediction but forgets to supply the feature extractor?

vinvinod replied:
I have a doubt here.
There is an option just to use the BoW if the feature extractor is not supplied. 
```
if (fe.get_num_dimensions() == 0) {
    temp = df.predict(extract_BoW_features(sentence));
} else {
    const std::vector<matrix<float, 0, 1> > &sent = sentence_to_feats(fe, sentence);
    temp = df.predict(extract_combined_features(sentence, sent));
}
```
So should it throw an error, or just give a warning that word-feature is not being used and just BoW is being used instead?

davisking replied:
The user should be required to use any model in the way it was trained.  If it was trained with a particular feature set then that feature set is definitely required when it is used or the results will be very inaccurate.

vinvinod replied:
We will have to store the fingerprint of the feature extractor when we are serializing the pure model. I will make the required changes and commit soon.

It won't be backward compatible! As the old pure model files wont be having the fingerprint of the the feature extractor.

davisking commented:
Yeah, I should have paid more attention to this when the first PR about
pure models was submitted. :(

Anyway, what you can do is add a different version string, the first string
serialized, and use that to identify pure models that have a fingerprint.
This way you could preserve backward compatibility which is pretty
important.

vinvinod commented:
I have made the required changes.
Now there is a version info saved along with the pure model file. We can increment this if there is a change in serialisation process in the future.

davisking commented:
Awesome.  Thanks for the PR :)

vinvinod commented:
I think you can close https://github.com/mit-nlp/MITIE/issues/36 now
